server/src/main/java/org/elasticsearch/cluster/routing/allocation/ShardWriteLoadDistributionMetrics.java

```diff
@@ -0,0 +1,275 @@
+import org.elasticsearch.telemetry.metric.MeterRegistry;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.atomic.AtomicReference;
+import java.util.concurrent.atomic.AtomicReferenceArray;
+import java.util.stream.IntStream;
+
+/**
+ * Publishes metrics about the distribution of shard write loads on each node in the cluster
+ */
+public class ShardWriteLoadDistributionMetrics {
+
+    private static final Logger logger = LogManager.getLogger(ShardWriteLoadDistributionMetrics.class);
+    private static final String WRITE_LOAD_DISTRIBUTION_METRIC_NAME = "es.allocator.shard_write_load.distribution.p%s.current";
+    public static final String WRITE_LOAD_PRIORITISATION_THRESHOLD_METRIC_NAME =
+        "es.allocator.shard_write_load.prioritisation_threshold.current";
+    public static final String WRITE_LOAD_PRIORITISATION_THRESHOLD_PERCENTILE_RANK_METRIC_NAME =
+        "es.allocator.shard_write_load.prioritisation_threshold.shard_count_exceeding.current";
+    public static final String WRITE_LOAD_SUM_METRIC_NAME = "es.allocator.shard_write_load.sum.current";
+    public static final Setting<Boolean> SHARD_WRITE_LOAD_METRICS_ENABLED_SETTING = Setting.boolSetting(
+        "cluster.routing.allocation.write_load_decider.shard_write_load_metrics.enabled",
+        true,
+        Setting.Property.Dynamic,
+        Setting.Property.NodeScope
+    );
+
+    private final DoubleHistogram shardWeightHistogram;
+    private final int[] percentiles;
+    private final ClusterService clusterService;
+    private final AtomicReferenceArray<List<DoubleWithAttributes>> lastWriteLoadDistributionValues;
+    private final AtomicReference<List<DoubleWithAttributes>> lastWriteLoadPrioritisationThresholdValues = new AtomicReference<>(List.of());
+    private final AtomicReference<List<LongWithAttributes>> lastShardCountExceedingPrioritisationThresholdValues = new AtomicReference<>(
+        List.of()
+    );
+    private final AtomicReference<List<DoubleWithAttributes>> lastWriteLoadSumValues = new AtomicReference<>(List.of());
+    private volatile boolean metricsEnabled = false;
+
+    public ShardWriteLoadDistributionMetrics(MeterRegistry meterRegistry, ClusterService clusterService) {
+        // 2 significant digits means error < 1% of any value in the range
+        this(meterRegistry, clusterService, 2, 50, 90, 95, 99, 100);
+    }
+
+    public ShardWriteLoadDistributionMetrics(
+        MeterRegistry meterRegistry,
+        ClusterService clusterService,
+        int numberOfSignificantDigits,
+        int... percentiles
+    ) {
+        this.clusterService = clusterService;
+        this.clusterService.getClusterSettings()
+            .initializeAndWatch(SHARD_WRITE_LOAD_METRICS_ENABLED_SETTING, value -> this.metricsEnabled = value);
+
+        // We can use ShortCountsHistogram because we don't expect any count to exceed Short.MAX_VALUE on a single node
```

Comment:
Should we be concerned about this? Short.MAX_VALUE appears to be ~32k. Particularly with the work around hollowing shards in serverless, I'm a little worried this expectation could be broken.